### ISSUE TITLE
SNOW-768234: execute() does not return Statement in bind mode

### ISSUE THREAD

**Issue by haggholm** [snowflakedb/snowflake-connector-nodejs#455]
Please answer these questions before submitting your issue. Thanks!

1. What version of NodeJS are you using (`node --version` and `npm --version`)?

```bash
$ node --version
v16.18.1
$ npm --version
8.16.0
```

2. What operating system and processor architecture are you using?

```bash
$  lsb_release -a 
No LSB modules are available.
Distributor ID:	Pop
Description:	Pop!_OS 22.04 LTS
Release:	22.04
Codename:	jammy
```

3. What are the component versions in the environment (`npm list`)?

(A lot of them, but they're not relevant.)

5. What did you do?

Execute a statement—any statement—with bind parameters. In our case, it looks something like

```typescript
const st = connection.execute({
  sqlText: query,
  streamResult: true,
  binds: parameters?.map((p) => p.value as Bind) ?? null,
  complete: onCompleteCallback,
);
const cancel = async () => new Promise((resolve, reject) => {
  st.cancel((err) => { /* ... */ });
});
this.cancellations.set(st.getRequestId(), cancel);
```

6. What did you expect to see?

For this to, well, work, but…:

7. What did you see instead?

Normally, `connection.execute()` returns a `Statement` object. However, if the number of parameters exceeds `snowflake-sdk`'s `ARRAY_BINDING_THRESHOLD`, `connection.execute()` doesn't return anything at all.

In the above case, obviously, we get an error like

```
TypeError: Cannot read properties of undefined (reading 'getRequestId')
```

I assume this is related to the fact that [in connection/statement.js](https://github.com/snowflakedb/snowflake-connector-nodejs/blob/9a8874b4b7185db0f8e5466e9c8f5b86cb864cf0/lib/connection/statement.js#L111), it has a return statement if there are no binds—but no return statement if there are!

```javascript
exports.createStatementPreExec = function (
  options, services, connectionConfig)
{
  // ...

  var isUsingBindUpload = false;
  // check array binding,
  if(numBinds > threshold)
  {
    // ...
  }

  if (!isUsingBindUpload)
  {
    return createRowStatementPreExec(
      options, context, services, connectionConfig);
  }
  // I'd have expected an else clause here!
};
```

8. Add this to get standard output.
 
(Not relevant)

**Comment by sfc-gh-dszmolka** [comment]
thank you for raising the issue with us and especially for providing so detailed description! the issue is easily reproducible with your hints - we'll take a look.

as a workaround if accessing the Statement's functions are a blocker issue for you (or the lack of access, in this situation), until this is is fixed, you can use the connection parameter `arrayBindingThreshold` which is available from `snowflake-sdk@1.6.17` and upwards; to bump the threshold which is used to determine whether the connector takes the bind upload path or not.
(default value is `100000`; a higher threshold will force the driver to take the old code path without bind uploader)

**Comment by sfc-gh-dszmolka** [comment]
driver team is looking into the issue still but in the meantime we got a possible workaround, which works both in bind upload mode and non bind upload mode:
```javascript
var insertRows = connection.execute({
    sqlText: 'insert into GH455.TEST.TEST (c1,c2,c3,c4,c5,c6,c7,c8,c9,c10,c11,c12,c13,c14,c15) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)',
     binds: [
..stuff..
        ],
    complete: function(err, stmt, rows) {
        if (err) {
            console.error('Failed to execute statement due to the following error: ' + err.message);
        } else {
            console.log(insertRows);
            //console.log('insertRows.getRequestId(): ' + insertRows.getRequestId());
            console.log('stmt.getRequestId(): ' + stmt.getRequestId());
            console.log('Number of rows produced: ' + rows.length);
        }
    }
});

```

getting `getRequestId()` (or other properties inside the statement) works from within the `stmt` inside the callback, in both modes. Perhaps this can be used as a workaround while we looking.

**Comment by sfc-gh-dszmolka** [comment]
https://github.com/snowflakedb/snowflake-connector-nodejs/pull/492

**Comment by sfc-gh-dszmolka** [comment]
issue should be now fixed with https://github.com/snowflakedb/snowflake-connector-nodejs/pull/492 , meaning it should now also return the `Statement` besides the above workaround.

it is readily available for use with npm i https://github.com/snowflakedb/snowflake-connector-nodejs.git & co. and I'll update this issue again when it's out with the next official release

**Comment by sfc-gh-dszmolka** [comment]
fix is out with release 1.6.23
